### PR TITLE
Use our own hashing architecture

### PR DESCRIPTION
**Pull request by lptr**
This replaces Guava's hashing. We need this for two reasons:

- it allows the addition of custom hash functions, like BLAKE2b which would not be possible with Guava's closed design
- it fixes a few performance issues that Guava only fixes in version 20 (which we can't upgrade to right now, because we still need to support Java 5 in places)

### CODE REVIEW

**Comment by lptr** [comment]
Branch build: https://builds.gradle.org/viewQueued.html?itemId=7781838

**Comment by lptr** [comment]
Branch build: https://builds.gradle.org/viewQueued.html?itemId=7789248

**Comment by lptr** [comment]
FTR, I found this, too: http://www.twmacinta.com/myjava/fast_md5.php, but it looks pretty irrelevant now. I wouldn't bother even measuring it.

**Comment by lptr** [comment]
Branch build: https://builds.gradle.org/viewQueued.html?itemId=7826837

**Comment by lptr** [comment]
Some interesting improvements:

| Test | Guava hashing | Our hashing | Diff |
| ---- | --------------- | ------------ | ---- |
| visiting zip trees | 3.386 s | 1.472 s | -56% |
| get IDE model on k9AndroidBuild for Android Studio | 245.5 ms | 232.5 ms | -5% |
| clean assemble on multiNative | 24.434 s | 22.246 s | -9% |
| **clean assemble on manyProjectsNative** | **13.296 s** | **6.276 s** | **-53%** |
| build with build file change on smallNativeMonolithic | 4.883 s | 4.358 s | -11% |

All results here: https://builds.gradle.org/viewLog.html?buildId=7827117&buildTypeId=Gradle_Check_PerformanceTestCoordinator&tab=report_project944_Performance

**Comment by wolfs** [comment]
Regarding the performance results: We are much faster for the native scenarios since your branch includes @ghale's changes ( #2656 ), while the baseline (4.2-20170817235727+0000) does not.
@ghale Should rebaseline all the native tests on master to lock in the improvement.

**Comment by wolfs** [comment]
I kicked off a performance build with this changes and always calculating the build cache key: https://builds.gradle.org/viewQueued.html?itemId=7838891&tab=queuedBuildOverviewTab
Let's see how the result for the fully up to date build changes.

File: subprojects/base-services/src/main/java/org/gradle/internal/hash/HashCode.java [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+        // Make sure hash codes are serializable with a single byte length
+        if (bytes.length < MIN_NUMBER_OF_BYTES || bytes.length > MAX_NUMBER_OF_BYTES) {
+            throw new IllegalArgumentException(String.format("Invalid hash code length: %d bytes", bytes.length));
+        }
+        return fromBytesNoCopy(bytes.clone());
+    }
+
+    public static HashCode fromInt(int value) {
+        return fromBytesNoCopy(new byte[] {
+            (byte) (value >> 24),
```
Comment (wolfs) [inline subprojects/base-services/src/main/java/org/gradle/internal/hash/HashCode.java]: Why did you change the order of the bytes in one of your commits?

File: subprojects/base-services/src/main/java/org/gradle/internal/hash/HashFunction.java [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.internal.hash;
+
+public interface HashFunction {
```
Comment (wolfs) [inline subprojects/base-services/src/main/java/org/gradle/internal/hash/HashFunction.java]: Javadoc would be nice.

File: subprojects/base-services/src/main/java/org/gradle/internal/hash/Hashing.java [hunk 3]
```diff
@@ -0,0 +1,206 @@
...
+package org.gradle.internal.hash;
+
+import com.google.common.base.Charsets;
+
+import java.nio.ByteBuffer;
+import java.nio.ByteOrder;
+import java.security.MessageDigest;
+import java.security.NoSuchAlgorithmException;
+
+public class Hashing {
```
Comment (wolfs) [inline subprojects/base-services/src/main/java/org/gradle/internal/hash/Hashing.java]: That is mostly another copied class. I think we should have some tests in place. Javadoc would be nice. Can we copy some of the tests from Guava, too?

File: subprojects/base-services/src/main/java/org/gradle/internal/hash/Hasher.java [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.internal.hash;
+
+public interface Hasher {
```
Comment (wolfs) [inline subprojects/base-services/src/main/java/org/gradle/internal/hash/Hasher.java]: Can we remove the unused methods from here (`putShort`, `putFloat`, `putChar`)?

File: subprojects/base-services/src/main/java/org/gradle/internal/hash/HashCode.java (hunk 1 above)
Comment (lptr) [reply subprojects/base-services/src/main/java/org/gradle/internal/hash/HashCode.java]: It makes the resulting value easier to interpret (as the bytes are not reversed). Since we only use `fromInt()` in tests, this is a valuable trait.

File: subprojects/base-services/src/main/java/org/gradle/internal/hash/Hasher.java (hunk 4 above)
Comment (lptr) [reply subprojects/base-services/src/main/java/org/gradle/internal/hash/Hasher.java]: I was on the fence about that. We can probably simplify more if we do that, so yeah.

**Comment by lptr** [comment]
About not cloning the bytes in the `BuildCacheHasher`: I don’t like that, and intentionally didn’t work around the cloning. A clone of 32 bytes for a short time should be very fast and the correctness is very important.

**Comment by wolfs** [comment]
Do we need to do something to not break the Kotlin DSL?

File: subprojects/base-services/src/testFixtures/groovy/org/gradle/internal/hash/TestFileHasher.groovy [hunk 5]
```diff
@@ -40,7 +36,7 @@ class TestFileHasher implements FileHasher {
 
     @Override
     HashCode hash(File file) {
-        return Files.hash(file, HASH_FUNCTION)
+        HASH_FUNCTION.hashBytes(file.bytes)
```
[the remark this answers is not in the document]
Comment (lptr) [reply subprojects/base-services/src/testFixtures/groovy/org/gradle/internal/hash/TestFileHasher.groovy]: Done

File: subprojects/internal-testing/src/main/groovy/org/gradle/test/fixtures/file/TestFile.java [hunk 6]
```diff
@@ -419,7 +419,7 @@ public TestFile assertIsDifferentFrom(TestFile other) {
 
     public String getMd5Hash() {
         try {
-            return Files.hash(this, Hashing.md5()).toString();
+            return Hashing.md5().hashBytes(Files.toByteArray(this)).toString();
```
[the remark this answers is not in the document]
Comment (lptr) [reply subprojects/internal-testing/src/main/groovy/org/gradle/test/fixtures/file/TestFile.java]: Done. We can't use `Files.hash()` because it expects a Guava `HashFunction`.

File: subprojects/base-services/src/jmh/java/org/gradle/internal/reflect/HashingAlgorithmsBenchmark.java [hunk 7]
```diff
@@ -0,0 +1,159 @@
...
+        .put("md5.bc", new MessageDigestHashProcessorFactory(new MD5.Digest()))
+        .put("sha1.java", new MessageDigestHashProcessorFactory(getDigest("SHA-1")))
+        .put("sha1.bc", new MessageDigestHashProcessorFactory(new SHA1.Digest()))
+        .put("blake2b.bc", new MessageDigestHashProcessorFactory(new Blake2b.Blake2b160()))
+        .put("murmur3.guava", new GuavaProcessorFactory(Hashing.murmur3_128()))
+        .build();
+
+    Random random = new Random(1234L);
+
+    // @Param({"16", "1024", "65536"})
```
[the remark this answers is not in the document]
Comment (lptr) [reply subprojects/base-services/src/jmh/java/org/gradle/internal/reflect/HashingAlgorithmsBenchmark.java]: Done

File: subprojects/base-services/src/main/java/org/gradle/internal/hash/Hashing.java (hunk 3 above)
Comment (lptr) [reply subprojects/base-services/src/main/java/org/gradle/internal/hash/Hashing.java]: Added a bit.

**Comment by lptr** [comment]
I don't see `HashValue` or `HashUtil` used in Kotlin DSL, but we'll know if we break it from the `master` builds I guess.

**Comment by lptr** [comment]
@wolfs can you take another look?

File: subprojects/resources/src/main/java/org/gradle/internal/resource/metadata/DefaultExternalResourceMetaData.java [hunk 8]
```diff
@@ -72,7 +72,7 @@ public String getEtag() {
         return etag;
     }
 
-    public HashValue getSha1() {
-        return sha1 == null ? null : HashValue.parse(sha1);
+    public HashCode getSha1() {
```
Comment (wolfs) [inline subprojects/resources/src/main/java/org/gradle/internal/resource/metadata/DefaultExternalResourceMetaData.java]: Shouldn't that be `@Nullable`? Or is that inherited from the base class?

File: subprojects/dependency-management/src/main/java/org/gradle/internal/resource/transfer/DefaultCacheAwareExternalResourceAccessor.java [hunk 9]
```diff
@@ -155,16 +155,16 @@ public ExternalResourceMetaData create() {
...
             ExternalResourceName sha1Location = location.append(".sha1");
             ExternalResource resource = delegate.resource(sha1Location, revalidate);
-            ExternalResourceReadResult<HashValue> result = resource.withContentIfPresent(new Transformer<HashValue, InputStream>() {
+            ExternalResourceReadResult<HashCode> result = resource.withContentIfPresent(new Transformer<HashCode, InputStream>() {
                 @Override
-                public HashValue transform(InputStream inputStream) {
+                public HashCode transform(InputStream inputStream) {
                     try {
                         String sha = IOUtils.toString(inputStream, "us-ascii");
-                        return HashValue.parse(sha);
```
Comment (wolfs) [inline subprojects/dependency-management/src/main/java/org/gradle/internal/resource/transfer/DefaultCacheAwareExternalResourceAccessor.java]: By not using `parse`, this is doing something differently now, right? Don't we need to keep the parse method?

File: subprojects/internal-testing/src/main/groovy/org/gradle/test/fixtures/file/TestFile.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (wolfs) [reply subprojects/internal-testing/src/main/groovy/org/gradle/test/fixtures/file/TestFile.java]: I would have gone "Full Guava" for hashing the file and getting the String, but it is better your way.

**Comment by wolfs** [comment]
@lptr I added some more comments.

File: subprojects/dependency-management/src/main/java/org/gradle/internal/resource/transfer/DefaultCacheAwareExternalResourceAccessor.java (hunk 9 above)
Comment (lptr) [reply subprojects/dependency-management/src/main/java/org/gradle/internal/resource/transfer/DefaultCacheAwareExternalResourceAccessor.java]: Just found this, too.

File: subprojects/dependency-management/src/main/java/org/gradle/internal/resource/transfer/DefaultCacheAwareExternalResourceAccessor.java [hunk 10]
```diff
@@ -155,16 +155,16 @@ public ExternalResourceMetaData create() {
...
+                        return HashUtil.parse(sha, 40);
```
Comment (wolfs) [inline subprojects/dependency-management/src/main/java/org/gradle/internal/resource/transfer/DefaultCacheAwareExternalResourceAccessor.java]: Why the padding to 40? I didn't see that in the original code?

Comment (lptr) [reply subprojects/dependency-management/src/main/java/org/gradle/internal/resource/transfer/DefaultCacheAwareExternalResourceAccessor.java]: It appears that we support SHA-1 hash codes that are shorter than 40 characters. We used to compare the values as `BigInteger`, which we don't do anymore. That's why we need the padding.

**Comment by lptr** [comment]
Restarted as #2817.
